src/main/java/org/kohsuke/github/GitHubBuilder.java

```diff
@@ -278,6 +279,11 @@ public GitHubBuilder withOAuthToken(String oauthToken, String user) {
         return this;
     }
 
+    public GitHubBuilder withCredentialProvider(final CredentialProvider credentialProvider) {
```

Comment:
I'd like to see all the other credential methods call this one. 

For example: 
```
    public GitHubBuilder withOAuthToken(String oauthToken, String user) {
        withCredentialProvider(ImmutableCredentialProvider.fromOauthToken(oauthToken));
        this.user = user;
        return this;
    }
```

That would let us remove the `oauthToken` and `password` instance fields from the class (and down the call chain).

Reply:
If we do this we lose the ability to know what kind of credentials we are using, and we will have to (I think) request the user to know if the current credentials are "user-aware" (GET "/user"). This will break a ton of tests that don't have the mapping for that call (not a big deal IMO), but also add another API call on GHBuilder#build(), and this does change semantics. We could potentially make it "lazy", so that we only lookup the user when it's actually requested, and not on setup, but I have to look into this.

Reply:
> If we do this we lose the ability to know what kind of credentials we are using

I feel like this isn't true.  There should be some way to do this. I might not be explaining myself well though.